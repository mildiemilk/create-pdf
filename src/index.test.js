const createDocDefinition = require('./index')

const mockData = {
  insured: {
    firstname: {
      value: 'AppMan',
    },
    lastname: {
      value: 'AgentMate',
    },
    gender: {
      value: {
        key: 'M',
        code: 'M',
      },
      dirty: true,
    },
    occupation: {
      value: {
        code: 'ก001004',
        key: 'ก001004',
        favorite: false,
      },
      dirty: true,
    },
    documentType: {
      value: {
        code: '1A',
        key: 'THAI_NATIONAL_ID_CARD',
      },
      dirty: true,
    },
    nationality: {
      value: {
        code: 'AF',
        key: 'AF',
      },
      dirty: true,
    },
    maritalStatus: {
      value: {
        code: '1',
        key: 'single',
      },
      dirty: true,
    },
    occupationJobDescription: {
      value: 'Test',
      dirty: true,
    }
  }
}

describe('createDocDefinition', () => {
  describe('When docDefinition template is styling', () => {
    const template = [{
      content: [
        'This is a standard paragraph, using default style',
        { text: 'insured.firstname.value', fontSize: 15 },
        {
          text: [
            'milk',
            { text: 'insured.lastname.value', fontSize: 15 },
            'than the rest.'
          ]
        }
      ],
      styles: {
        header: {
          fontSize: 22,
          bold: true
        },
        anotherStyle: {
          italics: true,
          alignment: 'right'
        }
      }
    }]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is columns', () => {
    const template = [
      {
        content: [
          'Just string',
          {
            columns: [
              {
                width: 'auto',
                text: 'insured.firstname.value'
              },
              {
                width: '*',
                text: 'Second column'
              },
              {
                // fixed width
                width: 100,
                text: 'insured.maritalStatus.value'
              },
              {
                // % width
                width: '20%',
                text: 'insured.nationality.value.code'
              }
            ],
            columnGap: 10
          },
          'milk eiei'
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is table', () => {
    const template = [
      {
        content: [
          {
            layout: 'insured.documentType.value.code',
            table: {
              headerRows: 1,
              widths: ['*', 'auto', 100, '*'],

              body: [
                ['First', 'Second', 'insured.firstname.value', 'The last one'],
                ['Value 1', 'insured.occupation.value.favorite', 33, 'Value 4'],
                [{ text: 'insured.lastname.value', bold: true }, 'Val 2', 'Val 3', 'insured.occupationJobDescription.value']
              ]
            }
          }
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is list', () => {
    const template = [
      {
        content: [
          'insured.firstname.value',
          {
            ul: [
              'Item 1',
              'Item 2',
              'insured.lastname.value',
              { text: 'insured.documentType.value.key', bold: true },
            ]
          },
          'Numbered list example:',
          {
            ol: [
              'Item 1',
              'Item 2',
              'insured.firstname.value'
            ]
          }
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is headers and footers', () => {
    const template = [
      {
        header: 'insured.lastname.value',
        footer: {
          columns: [
            'insured.firstname.value',
            { text: 'Right part', alignment: 'right' }
          ]
        },
        content: [
          'This is a standard paragraph, using default style',
          { text: 'insured.firstname.value', fontSize: 15 },
          {
            text: [
              'milk',
              { text: 'insured.lastname.value', fontSize: 15 },
              'than the rest.'
            ]
          }
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is stack of paragraphs', () => {
    const template = [
      {
        content: [
          'insured.firstname.value',
          'paragraph 2',
          {
            columns: [
              'first column is a simple text',
              'insured.firstname.value',
              [
                'paragraph A',
                'insured.firstname.value',
                'these paragraphs will be rendered one below another inside the column',
                [
                  'insured.firstname.value',
                  'milk'
                ],
              ]
            ]
          }
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is table of contents', () => {
    const template = [
      {
        content: [
          {
            toc: {
              title: { text: 'insured.firstname.value', style: 'header' }
            }
          },
          {
            text: 'insured.lastname.value',
            style: 'header',
            tocItem: true,
          }
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
  describe('When docDefinition template is encryption and access privileges', () => {
    const template = [
      {
        userPassword: 'insured.firstname.value',
        ownerPassword: '123456',
        permissions: {
          printing: 'highResolution',
          modifying: false,
          copying: false,
          annotating: 'insured.lastname.value',
          fillingForms: true,
          contentAccessibility: true,
          documentAssembly: true
        },
        content: [
          'milk',
          'insured.firstname.value'
        ]
      }
    ]
    it('Should map correctly', () => {
      const result = createDocDefinition(template, mockData)
      expect(result).toMatchSnapshot()
    })
  })
})