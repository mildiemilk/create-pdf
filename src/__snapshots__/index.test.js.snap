// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDocDefinition When docDefinition template is columns Should map correctly 1`] = `
Object {
  "content": Array [
    "Just string",
    Object {
      "columnGap": 10,
      "columns": Array [
        Object {
          "text": "AppMan",
          "width": "auto",
        },
        Object {
          "text": "Second column",
          "width": "*",
        },
        Object {
          "text": "1",
          "width": 100,
        },
        Object {
          "text": "AF",
          "width": "20%",
        },
      ],
    },
    "milk eiei",
  ],
}
`;

exports[`createDocDefinition When docDefinition template is encryption and access privileges Should map correctly 1`] = `
Object {
  "content": Array [
    "milk",
    "AppMan",
  ],
  "ownerPassword": "123456",
  "permissions": Object {
    "annotating": "AgentMate",
    "contentAccessibility": true,
    "copying": false,
    "documentAssembly": true,
    "fillingForms": true,
    "modifying": false,
    "printing": "highResolution",
  },
  "userPassword": "AppMan",
}
`;

exports[`createDocDefinition When docDefinition template is headers and footers Should map correctly 1`] = `
Object {
  "content": Array [
    "This is a standard paragraph, using default style",
    Object {
      "fontSize": 15,
      "text": "AppMan",
    },
    Object {
      "text": Array [
        "milk",
        Object {
          "fontSize": 15,
          "text": "AgentMate",
        },
        "than the rest.",
      ],
    },
  ],
  "footer": Object {
    "columns": Array [
      "AppMan",
      Object {
        "alignment": "right",
        "text": "Right part",
      },
    ],
  },
  "header": "AgentMate",
}
`;

exports[`createDocDefinition When docDefinition template is image Should map correctly 1`] = `
Object {
  "content": Array [
    Object {
      "image": undefined,
      "text": "milk",
    },
  ],
}
`;

exports[`createDocDefinition When docDefinition template is list Should map correctly 1`] = `
Object {
  "content": Array [
    "AppMan",
    Object {
      "ul": Array [
        "Item 1",
        "Item 2",
        "AgentMate",
        Object {
          "bold": true,
          "text": "THAI_NATIONAL_ID_CARD",
        },
      ],
    },
    "Numbered list example:",
    Object {
      "ol": Array [
        "Item 1",
        "Item 2",
        "AppMan",
      ],
    },
  ],
}
`;

exports[`createDocDefinition When docDefinition template is stack of paragraphs Should map correctly 1`] = `
Object {
  "content": Array [
    "AppMan",
    "paragraph 2",
    Object {
      "columns": Array [
        "first column is a simple text",
        "AppMan",
        Array [
          "paragraph A",
          "AppMan",
          "these paragraphs will be rendered one below another inside the column",
          Array [
            "AppMan",
            "milk",
          ],
        ],
      ],
    },
  ],
}
`;

exports[`createDocDefinition When docDefinition template is styling Should map correctly 1`] = `
Object {
  "content": Array [
    "This is a standard paragraph, using default style",
    Object {
      "fontSize": 15,
      "text": "AppMan",
    },
    Object {
      "text": Array [
        "milk",
        Object {
          "fontSize": 15,
          "text": "AgentMate",
        },
        "than the rest.",
      ],
    },
  ],
  "styles": Object {
    "anotherStyle": Object {
      "alignment": "right",
      "italics": true,
    },
    "header": Object {
      "bold": true,
      "fontSize": 22,
    },
  },
}
`;

exports[`createDocDefinition When docDefinition template is table Should map correctly 1`] = `
Object {
  "content": Array [
    Object {
      "layout": "1A",
      "table": Object {
        "body": Array [
          Array [
            "First",
            "Second",
            "AppMan",
            "The last one",
          ],
          Array [
            "Value 1",
            false,
            33,
            "Value 4",
          ],
          Array [
            Object {
              "bold": true,
              "text": "AgentMate",
            },
            "Val 2",
            "Val 3",
            "Test",
          ],
        ],
        "headerRows": 1,
        "widths": Array [
          "*",
          "auto",
          100,
          "*",
        ],
      },
    },
  ],
}
`;

exports[`createDocDefinition When docDefinition template is table of contents Should map correctly 1`] = `
Object {
  "content": Array [
    Object {
      "toc": Object {
        "title": Object {
          "style": "header",
          "text": "AppMan",
        },
      },
    },
    Object {
      "style": "header",
      "text": "AgentMate",
      "tocItem": true,
    },
  ],
}
`;
